models/vc/FreeVC/data.py

```diff
@@ -0,0 +1,284 @@
+from .mel_processing import spectrogram_torch
+from .commons import rand_spec_segments, slice_segments
+
+import os
+import random
+import torch
+from torch.utils.data import Dataset, Sampler
+import torchaudio
+
+
+def read_txt_lines(path):
+    ans = []
+    with open(str(path), "r") as f:
+        for line in f.readlines():
+            line = line.strip()
+            if line != "":
+                ans.append(line)
+    return ans
+
+
+class FreeVCDataset(Dataset):
+    def __init__(self, audiopaths, hparams):
+        self.audiopaths = read_txt_lines(audiopaths)
+
+        self.sampling_rate = hparams.data.sampling_rate
+        self.filter_length = hparams.data.filter_length
+        self.hop_length = hparams.data.hop_length
+        self.win_length = hparams.data.win_length
+        self.sampling_rate = hparams.data.sampling_rate
+        self.use_sr = hparams.train.use_sr
+        self.use_spk = hparams.model.use_spk
+        self.spec_len = hparams.train.max_speclen
+
+        self.vctk_16k_dir = hparams.preprocess.vctk_16k_dir
+        self.spk_dir = hparams.preprocess.spk_dir
+        self.ssl_dir = hparams.preprocess.ssl_dir
+        self.sr_dir = hparams.preprocess.sr_dir
+
+        random.shuffle(self.audiopaths)
+        self._filter()
+
+    def _filter(self):
+        """
+        Filter text & store spec lengths
+        """
+        # Store spectrogram lengths for Bucketing
+        # wav_length ~= file_size / (wav_channels * Bytes per dim) = file_size / (1 * 2)
+        # spec_length = wav_length // hop_length
+
+        lengths = []
+        for audiopath in self.audiopaths:
+            path = os.path.join(self.vctk_16k_dir, audiopath)
+            lengths.append(os.path.getsize(path) // (2 * self.hop_length))
+        self.lengths = lengths
+
+    @torch.no_grad()
+    def load_sample(self, filename):
+        filepath = os.path.join(self.vctk_16k_dir, filename)
```

Comment:
Why is this line hard-coded? Is it possible to select datasets in config?

Reply:
The original implementation only trains on VCTK dataset. 

+ Data preprocessing uses the file structure of VCTK dataset to retrieve speaker tags.
+ When splitting train/val/test set, every speaker's samples are split randomly. It ensures that every speaker is in train & val & test set.

It's possible to support other datasets if we can perform the same operations on them.

Reply:
Thank you for your explanation! 
@RMSnow For this implementation, do we expect a universal model that can be trained on any dataset?

Reply:
Yes, a universal FreeVC for any datasets are welcome. I think only the FreeVC's model part need to be integrated in Amphion.